src/js/components/Util/CustomTable.tsx

```diff
@@ -0,0 +1,128 @@
+import { useTranslationFn } from '@/hooks';
+
+export interface CustomTableColumn<T> extends TableColumnType<T> {
+  isEmpty?: (value: any, record?: T) => boolean; // eslint-disable-line @typescript-eslint/no-explicit-any
+  isEmptyDefaultCheck?: boolean;
+}
+
+export type CustomTableColumns<T> = CustomTableColumn<T>[];
+
+type VisibilityFunc<T> = (record: T) => boolean;
+type RowSelectorFunc<T> = (record: T) => string;
+
+export const addVisibilityProperty = <T,>(r: T[], visibilityFn: VisibilityFunc<T>): WithVisible<T>[] => {
+  return r.map((item) => ({
+    ...item,
+    isVisible: visibilityFn(item),
+  }));
+};
+
+interface CustomTableProps<T> {
+  dataSource: T[];
+  columns: CustomTableColumns<T>;
+  rowKey: TableProps<WithVisible<T>>['rowKey'];
+  isDataKeyVisible: VisibilityFunc<T>;
+  expandedRowRender?: (record: T) => React.ReactNode;
+}
+
+const CustomTable = <T,>({ dataSource, columns, rowKey, isDataKeyVisible, expandedRowRender }: CustomTableProps<T>) => {
+  type VT = WithVisible<T>;
+
+  const [searchParams, setSearchParams] = useSearchParams();
+  const urlExpanded = useMemo(
+    () => searchParams.get(EXPANDED_QUERY_PARAM_KEY)?.split(',').filter(Boolean) || [],
+    [searchParams]
+  );
+  const [expandedRowKeys, setExpandedRowKeys] = useState<string[]>([]);
+  const api = useNotify();
+  const t = useTranslationFn();
+
+  const getKey = useMemo<RowSelectorFunc<VT>>(() => {
+    if (typeof rowKey === 'function') {
+      return rowKey as RowSelectorFunc<VT>;
+    }
+    return (record: VT) => String(record[rowKey as keyof VT]);
+  }, [rowKey]);
+
+  const updatedColumns = useTranslatedTableColumnTitles<T>(columns || []) as CustomTableColumns<VT>;
+  const dataSourceWithVisibility = addVisibilityProperty<T>(dataSource, isDataKeyVisible);
+  const updatedColumnsWithVisibility = updatedColumns!.map((col) => {
+    if (col.isEmpty) {
+      return {
+        ...col,
+        hidden: !dataSourceWithVisibility.some((record) => !col.isEmpty!(record[col.dataIndex as keyof T], record)),
+      };
+    } else if (col.isEmptyDefaultCheck) {
+      return {
+        ...col,
+        hidden: !dataSourceWithVisibility.some((record) => record[col.dataIndex as keyof T]),
+      };
+    }
+    return col;
+  });
+
+  const validKeys = useMemo(() => {
+    const keysSet = new Set(dataSourceWithVisibility.filter(({ isVisible }) => isVisible).map(getKey));
+    return urlExpanded.filter((key) => keysSet.has(key));
+  }, [dataSourceWithVisibility, getKey, urlExpanded]);
+
+  useEffect(() => {
+    setExpandedRowKeys(validKeys);
+    if (validKeys.length !== urlExpanded.length) {
+      api.warning({
+        message: t('table.invalid_row_keys_title'),
+        description: t('table.invalid_row_keys_description'),
+      });
+
+      if (validKeys.length) {
+        searchParams.set(EXPANDED_QUERY_PARAM_KEY, validKeys.join(','));
+      } else {
+        searchParams.delete(EXPANDED_QUERY_PARAM_KEY);
```

Comment:
I don't believe mutating a state variable like this is a valid thing to do

Reply:
I am redoing the full component, there were some bugs in there